apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -0,0 +1,169 @@
+  //   x: the x attribute found in <block/> element
+  //   y: the y attribute found in <block/> element
+  const xmlBlocks = Blockly.Xml.domToBlockSpace(tempWorkspace, xml);
+
+  const stateToLoad = Blockly.serialization.workspaces.save(tempWorkspace);
+
+  // Create a map of ids (key) and blocks (value).
+  const blockIdMap = stateToLoad.blocks.blocks.reduce(
+    (map, block) => map.set(block.id, block),
+    new Map()
+  );
+
+  addPositionsToState(xmlBlocks, blockIdMap);
+  tempWorkspace.dispose();
+  return stateToLoad;
+}
+
+/**
+ * Adds x/y values from XML to JSON serialization.
+ * @param {Array<Object>} xmlBlocks - an array of "block" objects containing a block and x/y coordinates
+ * @param {Map<String, Object>} blockIdMap - a map of ids (keys) and blocks (values)
+ */
+function addPositionsToState(xmlBlocks, blockIdMap) {
+  xmlBlocks.forEach(xmlBlock => {
+    const block = blockIdMap.get(xmlBlock.blockly_block.id);
+    if (block) {
+      // Do not change block values if xmlBlock values are NaN (unspecified in XML)
+      block.x = xmlBlock.x || block.x;
+      block.y = xmlBlock.y || block.y;
+    }
+  });
+}
+
+/**
+ * Position blocks on a workspace (if they do not already have positions)
+ * @param {Blockly.Workspace} workspace - the current Blockly workspace
+ * @param {function} positionBlock - moves a single block using the current cursor coordinates
+ * @param {Array<block>} blocks - an array of block objects
+ */
+export function positionBlocksOnWorkspace(
+  workspace,
+  positionBlock,
+  blocks = workspace.getTopBlocks(SORT_BY_POSITION)
+) {
+  if (!workspace.rendered) {
+    return blocks;
+  }
+  const metrics = workspace.getMetrics() || {contentWidth: 0, viewWidth: 0};
+  const contentWidth = metrics.contentWidth;
+  const viewWidth = metrics.viewWidth;
+  const padding = viewWidth ? 16 : 0;
+
+  // The "cursor" object tracks a position on the workspace that starts in the top
+  // corner, then moves below each block that is manually repositioned.
+  let cursor = {
+    x: padding,
+    y: padding,
+  };
+  if (workspace.RTL && viewWidth) {
+    // Position the cursor from the right of the workspace.
+    cursor.x = viewWidth - padding;
+  } else if (workspace.RTL && !viewWidth) {
+    // Position the cursor from the right of the block.
+    cursor.x = contentWidth - padding;
+  }
+
+  blocks.forEach(block => positionBlock(block, workspace, cursor));
+}
+
+/**
+ * Use a cursor to position a block on a workspace (if it does not already have a position)
+ * @param {Blockly.Block} block - the block to be moved
+ * @param {Blockly.Workspace} workspace - the current Blockly workspace
+ * @param {object} cursor - a location for moving a block, if needed
+ * @param {number} cursor.x - an x-coordinate for moving a block
+ * @param {number} cursor.y - a y-coordinate for moving a block
+ */
+export function positionBlock(block, workspace, cursor) {
+  const verticalSpaceBetweenBlocks = 10;
```

Comment:
Now that we have two versions of `positionBlocks`, would it make sense to create constants at the top of the file for these duplicates? Same with the `frameSvgTop`, etc below.

Reply:
Yes, definitely. I should be able to move any constants that aren’t based on workspace metrics.